### ISSUE TITLE
HermesMockExpect#avroMessagesOnTopic with a predicate does not work (when not all messages matches the predicate)

### ISSUE THREAD

**Issue by platan** [allegro/hermes#1152]
https://github.com/allegro/hermes/blob/0a512c8b0fe86cfca8e50e347f4b5ce2c8e37686/hermes-mock/src/main/java/pl/allegro/tech/hermes/mock/HermesMockExpect.java#L52
does not work when not all messages matches the predicate. 

An example:
```groovy
def "should get some filtered messages as avro"() {
        given:
        def topicName = "get-some-filtered-avro-topic"
        hermes.define().avroTopic(topicName)

        def messages = (1..5).collect { new TestMessage("key-" + it, "value-" + it) }
        def filter = { TestMessage m -> m.key.startsWith("key-1") || m.key.startsWith("key-3") }

        when:
        messages.each { publish(topicName, it) }

        then:
        hermes.expect().avroMessagesOnTopic(topicName, 2, schema, TestMessage, filter)
    }
```
I expect it to pass, but I get a failure instead:
`Expected exactly 2 requests matching the following pattern but received 5`

**Comment by dankraw** [comment]
Hmm, I've looked at the code and I think it's not an error. 

All of the messages were published with `messages.each { publish(topicName, it) }`, so hermes received 5 messages.
The additional thing is, you can look at the received messages from hermes perspective and apply the predicate to them (like check if the deserialized body contained `key-` prefix on all of them).

It is not a predicate that is applied on client side nor hermes-frontends side (there is no such thing as filters on hermes-frontends when accepting messages - if the message can be read using topic's schema, it's fine).

@cristaloleg could you confirm if what I say is correct?

**Comment by cristaloleg** [comment]
So...it's not an expected behaviour. In Hermes Mock filtering is a tool to let user check what events have appeared on a topic, but due to bad predicate in tests  it wasn't covered properly.

Fixed in #1153

**Comment by dankraw** [comment]
Seems logical, thanks @cristaloleg
